File: src/fr/geobert/radis/ui/drawer/NavDrawerListAdapter.java
package fr.geobert.radis.ui.drawer;

import android.app.Activity;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import fr.geobert.radis.R;

import java.util.ArrayList;

public class NavDrawerListAdapter extends BaseAdapter {
    private Context context;
    private ArrayList<NavDrawerItem> navDrawerItems;

    private class ViewHolder {
        ImageView icon;
        TextView header;
        TextView title;
        LinearLayout headerCont;
        LinearLayout itemCont;

        ViewHolder(View v) {
            icon = (ImageView) v.findViewById(R.id.drawer_item_img);
            title = (TextView) v.findViewById(R.id.drawer_item_lbl);
            header = (TextView) v.findViewById(R.id.drawer_header_lbl);
            headerCont = (LinearLayout) v.findViewById(R.id.drawer_header);
            itemCont = (LinearLayout) v.findViewById(R.id.drawer_item);
        }
    }

    public NavDrawerListAdapter(Context context, ArrayList<NavDrawerItem> navDrawerItems) {
        super();
        this.context = context;
        this.navDrawerItems = navDrawerItems;
    }

    @Override
    public int getCount() {
        return navDrawerItems.size();
    }

    @Override
    public Object getItem(int i) {
        return navDrawerItems.get(i);
    }

    @Override
    public long getItemId(int i) {
        return i;
    }

    @Override
    public View getView(int i, View view, ViewGroup viewGroup) {
        ViewHolder h;
        if (view == null) {
            LayoutInflater l = (LayoutInflater) context.getSystemService(Activity.LAYOUT_INFLATER_SERVICE);
            view = l.inflate(R.layout.drawer_item, null);
            h = new ViewHolder(view);
            view.setTag(h);
        } else {
            h = (ViewHolder) view.getTag();
        }

        NavDrawerItem item = navDrawerItems.get(i);
        if (item.isHeader()) {
            h.itemCont.setVisibility(View.GONE);
            h.headerCont.setVisibility(View.VISIBLE);
            h.header.setText(item.getTitle());
        } else {
            h.itemCont.setVisibility(View.VISIBLE);
            h.headerCont.setVisibility(View.GONE);
            h.title.setText(item.getTitle());
            if (item.getIcon() != 0) {
                h.icon.setImageResource(item.getIcon());
            }
        }
        return view;
    }
}

File: app/src/main/kotlin/fr/geobert/radis/ui/drawer/NavDrawerListAdapter.kt
package fr.geobert.radis.ui.drawer

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import fr.geobert.radis.R
import java.util.*

class NavDrawerListAdapter(private val context: Context, private val navDrawerItems: ArrayList<NavDrawerItem>) : BaseAdapter() {

    private inner class ViewHolder internal constructor(v: View) {
        internal var icon: ImageView
        internal var header: TextView
        internal var title: TextView
        internal var headerCont: LinearLayout
        internal var itemCont: LinearLayout

        init {
            icon = v.findViewById(R.id.drawer_item_img) as ImageView
            title = v.findViewById(R.id.drawer_item_lbl) as TextView
            header = v.findViewById(R.id.drawer_header_lbl) as TextView
            headerCont = v.findViewById(R.id.drawer_header) as LinearLayout
            itemCont = v.findViewById(R.id.drawer_item) as LinearLayout
        }
    }

    override fun getCount(): Int {
        return navDrawerItems.size
    }

    override fun getItem(i: Int): Any {
        return navDrawerItems.get(i)
    }

    override fun isEnabled(position: Int): Boolean {
        return !(getItem(position) as NavDrawerItem).isHeader
    }

    override fun getItemId(i: Int): Long {
        return i.toLong()
    }

    override fun getView(i: Int, v: View?, viewGroup: ViewGroup): View {
        var view = v
        val h: ViewHolder
        if (view == null) {
            val l = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            view = l.inflate(R.layout.drawer_item, null, false)
            h = ViewHolder(view)
            view?.tag = h
        } else {
            h = view.tag as ViewHolder
        }

        val item = navDrawerItems.get(i)
        if (item.isHeader) {
            h.itemCont.visibility = View.GONE
            h.headerCont.visibility = View.VISIBLE
            h.header.text = item.title
        } else {
            h.itemCont.visibility = View.VISIBLE
            h.headerCont.visibility = View.GONE
            h.title.text = item.title
            if (item.icon != 0) {
                h.icon.setImageResource(item.icon)
                h.icon.visibility = View.VISIBLE
            } else {
                h.icon.visibility = View.GONE
            }
        }
        return view as View
    }
}
